Services/Authentication/classes/class.ilSession.php

```diff
@@ -143,54 +138,66 @@ public static function _writeData(string $a_session_id, string $a_data): bool
         $now = time();
 
         // prepare session data
-        $fields = array(
-            "user_id" => array("integer", (int) (self::get('_authsession_user_id') ?? 0)),
-            "expires" => array("integer", self::getExpireValue()),
-            "data" => array("clob", $a_data),
-            "ctime" => array("integer", $now),
-            "type" => array("integer", (int) (self::get("SessionType") ?? 0))
-            );
-        if ($ilClientIniFile->readVariable("session", "save_ip")) {
-            $fields["remote_addr"] = array("text", $_SERVER["REMOTE_ADDR"]);
+        $fields = [
+            'user_id' => [ilDBConstants::T_INTEGER, (int) (self::get('_authsession_user_id') ?? 0)],
+            'expires' => [ilDBConstants::T_INTEGER, self::getExpireValue()],
+            'data' => [ilDBConstants::T_CLOB, $a_data],
+            'ctime' => [ilDBConstants::T_INTEGER, $now],
+            'type' => [ilDBConstants::T_INTEGER, (int) (self::get('SessionType') ?? 0)]
+        ];
+        if ($ilClientIniFile->readVariable('session', 'save_ip')) {
+            $fields['remote_addr'] = [ilDBConstants::T_TEXT, $_SERVER['REMOTE_ADDR'] ?? ''];
         }
 
         if (self::_exists($a_session_id)) {
             // note that we do this only when inserting the new record
             // updating may get us other contexts for the same session, especially ilContextWAC, which we do not want
-            if (class_exists("ilContext") && ilContext::isSessionMainContext()) {
-                $fields["context"] = array("text", ilContext::getType());
+            if (class_exists('ilContext') && ilContext::isSessionMainContext()) {
+                $fields['context'] = [ilDBConstants::T_TEXT, ilContext::getType()];
             }
             $ilDB->update(
-                "usr_session",
+                'usr_session',
                 $fields,
-                array("session_id" => array("text", $a_session_id))
+                ['session_id' => [ilDBConstants::T_TEXT, $a_session_id]]
             );
         } else {
-            $fields["session_id"] = array("text", $a_session_id);
-            $fields["createtime"] = array("integer", $now);
+            $fields['session_id'] = [ilDBConstants::T_TEXT, $a_session_id];
+            $fields['createtime'] = [ilDBConstants::T_INTEGER, $now];
 
             // note that we do this only when inserting the new record
             // updating may get us other contexts for the same session, especially ilContextWAC, which we do not want
-            if (class_exists("ilContext")) {
-                $fields["context"] = array("text", ilContext::getType());
+            if (class_exists('ilContext')) {
+                $fields['context'] = [ilDBConstants::T_TEXT, ilContext::getType()];
             }
 
-            $ilDB->insert("usr_session", $fields);
+            try {
+                $ilDB->insert('usr_session', $fields);
+            } catch (PDOException $e) {
+                if ($e->getCode() === 23000) {
```

Comment:
`23000` is the `SQLState` code for a duplicate key constraint violation. Of course this could be extracted to a PHP constant.

> Error number: 1022; Symbol: [ER_DUP_KEY](https://dev.mysql.com/doc/mysql-errors/8.0/en/server-error-reference.html#error_er_dup_key); SQLSTATE: 23000